[PyTorch] Update expand_outplace API to match expand_inplace

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #55351 [PyTorch] Fix const correctness for resize native functions
* #55685 [PyTorch] Avoid double indirection in MaybeOwned's borrowed state
* #55684 [PyTorch] Change MaybeOwned tests to use intrusive_ptr and Tensor
* #55553 [PyTorch] Mark borrowed case as C10_LIKELY in MaybeOwned
* #55419 [PyTorch] Allow copy operations on MaybeOwned
* #55316 [PyTorch] Add & use inferExpandGeometry_dimvector
* #55247 [PyTorch][easy] Use sizes()[x] instead of size(x) in addr
* #55246 [PyTorch] Update expand_size API to match expand_inplace
* **#55245 [PyTorch] Update expand_outplace API to match expand_inplace**
* #55244 [PyTorch] Add MaybeOwned::operator*() &&
* #55180 [PyTorch] Use infer_size_dimvector in ExpandUtils
* #55065 [PyTorch] New expand_inplace API with MaybeOwned<Tensor> and no unary tuples

Like `expand_inplace`, `expand_outplace` now returns
`MaybeOwned<Tensor>` in most cases. I wasn't confident around the
ownership semantics of the `TensorList` -> `std::vector<Tensor>` case, so I
left that one alone.

Differential Revision: [D27522811](https://our.internmc.facebook.com/intern/diff/D27522811/)

swolchok committed:
[PyTorch] Update expand_outplace API to match expand_inplace

Like `expand_inplace`, `expand_outplace` now returns
`MaybeOwned<Tensor>` in most cases. I wasn't confident around the
ownership semantics of the `TensorList` -> `std::vector<Tensor>` case, so I
left that one alone.

Differential Revision: [D27522811](https://our.internmc.facebook.com/intern/diff/D27522811/)

[ghstack-poisoned]

swolchok committed:
rebase, fix typo in D27522811 on "[PyTorch] Update expand_outplace API to match expand_inplace"


Like `expand_inplace`, `expand_outplace` now returns
`MaybeOwned<Tensor>` in most cases. I wasn't confident around the
ownership semantics of the `TensorList` -> `std::vector<Tensor>` case, so I
left that one alone.

Differential Revision: [D27522811](https://our.internmc.facebook.com/intern/diff/D27522811/)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/ExpandUtils.h:
```diff
@@ -103,46 +103,128 @@ inline std::tuple<c10::MaybeOwned<Tensor>, c10::MaybeOwned<Tensor>> expand_inpla
+inline std::tuple<c10::MaybeOwned<Tensor>, c10::MaybeOwned<Tensor>>
+expand_outplace(const Tensor &to_expand1, Tensor &&to_expand2) = delete;
+inline std::tuple<c10::MaybeOwned<Tensor>, c10::MaybeOwned<Tensor>>
+expand_outplace(Tensor &&to_expand1, Tensor &&to_expand2) = delete;
```
I wonder if there is anything we could do with implicit conversions to avoid having to explicitly delete all combinatorial combinations of `Tensor&&` overloads, cuz it kind of bytes having a two argument function balloon into four variants here.

ezyang replied:
And it's even worse for three argument functions

swolchok replied:
Implicit conversions are not quite good enough. I tried adding the following helper:

```
// NonRvalueArgument<T> can be used instead of const T& as an argument
// type if you wish to disallow rvalues (e.g., a prvalue temporary)
// from being passed. The advantage is that you avoid having to
// explicitly delete 2**N - 1 overloads given N such arguments.
template <typename T>
class NonRvalueArgument {
  const T& ref;
 public:
  /* implicit */ NonRvalueArgument(const T& r) : ref(r) {}
  NonRvalueArgument(T&& r) = delete;

  operator const T&() const {
    return ref;
  }
};
```

But you cannot call `Tensor` methods on a `NonRvalueArgument<Tensor>`. We would have to implement `operator*` and `operator->` and use pointer syntax. If you would prefer that, I can send out a diff to do it, but either way there is something ugly involved, and with the deleted overloads they're at least not cluttering the function body.

ezyang replied:
Well, at least pointer syntax doesn't require 2**N overloads. NBD, you can land this as is, but if you ever hit a 4-ary probably crack this one open.

swolchok committed:
see if OSS doctest failures about non-ASCII character go away on "[PyTorch] Update expand_outplace API to match expand_inplace"


Like `expand_inplace`, `expand_outplace` now returns
`MaybeOwned<Tensor>` in most cases. I wasn't confident around the
ownership semantics of the `TensorList` -> `std::vector<Tensor>` case, so I
left that one alone.

Differential Revision: [D27522811](https://our.internmc.facebook.com/intern/diff/D27522811/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Update expand_outplace API to match expand_inplace"


Like `expand_inplace`, `expand_outplace` now returns
`MaybeOwned<Tensor>` in most cases. I wasn't confident around the
ownership semantics of the `TensorList` -> `std::vector<Tensor>` case, so I
left that one alone.

Differential Revision: [D27522811](https://our.internmc.facebook.com/intern/diff/D27522811/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Update expand_outplace API to match expand_inplace"


Like `expand_inplace`, `expand_outplace` now returns
`MaybeOwned<Tensor>` in most cases. I wasn't confident around the
ownership semantics of the `TensorList` -> `std::vector<Tensor>` case, so I
left that one alone.

Differential Revision: [D27522811](https://our.internmc.facebook.com/intern/diff/D27522811/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Update expand_outplace API to match expand_inplace"


Like `expand_inplace`, `expand_outplace` now returns
`MaybeOwned<Tensor>` in most cases. I wasn't confident around the
ownership semantics of the `TensorList` -> `std::vector<Tensor>` case, so I
left that one alone.

Differential Revision: [D27522811](https://our.internmc.facebook.com/intern/diff/D27522811/)

[ghstack-poisoned]
